runelite-client/src/main/java/net/runelite/client/game/chatbox/ChatboxTextInput.java

```diff
@@ -0,0 +1,663 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.game.chatbox;
+
+import com.google.inject.Inject;
+import java.awt.Rectangle;
+import java.awt.Toolkit;
+import java.awt.datatransfer.DataFlavor;
+import java.awt.datatransfer.StringSelection;
+import java.awt.datatransfer.UnsupportedFlavorException;
+import java.awt.event.KeyEvent;
+import java.awt.event.MouseEvent;
+import java.io.IOException;
+import java.util.function.Consumer;
+import java.util.function.IntPredicate;
+import java.util.function.Predicate;
+import java.util.function.ToIntFunction;
+import javax.swing.SwingUtilities;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import net.runelite.api.FontTypeFace;
+import net.runelite.api.FontID;
+import net.runelite.api.WidgetType;
+import net.runelite.api.widgets.JavaScriptCallback;
+import net.runelite.api.widgets.Widget;
+import net.runelite.api.widgets.WidgetPositionMode;
+import net.runelite.api.widgets.WidgetSizeMode;
+import net.runelite.api.widgets.WidgetTextAlignment;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.input.KeyListener;
+import net.runelite.client.input.MouseListener;
+
+@Slf4j
+public class ChatboxTextInput extends ChatboxInput implements KeyListener, MouseListener
+{
+	private static final int CURSOR_FLASH_RATE_MILLIS = 1000;
+
+	private final ChatboxPanelManager chatboxPanelManager;
+	private final ClientThread clientThread;
+
+	private static IntPredicate getDefaultCharValidator()
+	{
+		return i -> i >= 32 && i < 127;
+	}
+
+	@Getter
+	private String prompt;
+
+	private StringBuffer value = new StringBuffer();
+
+	@Getter
+	private int cursor = 0;
+
+	@Getter
+	private int cursorEnd = 0;
+
+	private int selectionStart = -1;
+	private int selectionEnd = -1;
+
+	@Getter
+	private IntPredicate charValidator = getDefaultCharValidator();
+
+	@Getter
+	private Runnable onClose = null;
+
+	@Getter
+	private Consumer<String> onDone = null;
+
+	@Getter
+	private Consumer<String> onChanged = null;
+
+	@Getter
+	private int fontID = FontID.QUILL_1;
+
+	// This is a lambda so I can have atomic updates for it's captures
+	private ToIntFunction<MouseEvent> getCharOffset = null;
+	private Predicate<MouseEvent> isInBounds = null;
+
+	private boolean built = false;
+
+	@Inject
+	protected ChatboxTextInput(ChatboxPanelManager chatboxPanelManager, ClientThread clientThread)
```

Comment:
private again

Reply:
This class is designed to be subclassed

Reply:
I do not see where it is subclassed. And if that is the case, maybe it should not use injection

Reply:
It is not subclassed, it is __designed__ to be subclassed if in the future we want more complicated input boxes. I can manually inject it, but that seems several times worse than just letting guice do it.

Reply:
Yeah, I just dont think that designing classes that are supposed to be created through Guice to be subclassed in future is good idea. The constructor has only like 2 fields anyway, so creating it without Guice might be better.

Reply:
I don't see the advantage to adding redundant boilerplate.

Reply:
The advantage is that it will not create more technical debt for this issue: https://github.com/runelite/runelite/issues/617

Reply:
I don't see how this has anything to do with #617.

Reply:
From what I've understood from their wiki pages is that if Guice is managing the class, you are not supposed to call the constructor unless you are unit testing the class

Reply:
Why are you not supposed to?

Reply:
To prevent accidental misuse of the classes. In this case it is not so easy I guess, but the most common case where people do stupid things with non-private ctors for Guice classes is that they try to create the class by themselves and `@Inject` to it at same time, in what case the injected fields are null.

Reply:
If you have no `@Inject` fields then you can't have that problem. I don't see how calling a `@Inject` constructor is worse than calling the exact same method without `@Inject`.

Reply:
Yeah it is mostly only problem with `@Inject` fields. I guess this is fine then